api/src/main/java/org/pmiops/workbench/workspaces/WorkspaceServiceImpl.java

```diff
@@ -495,4 +498,122 @@ public Workspace setPublished(
 
     return this.saveWithLastModified(workspace);
   }
+
+  @Override
+  public List<UserRecentWorkspace> getRecentWorkspaces() {
+    long userId = userProvider.get().getUserId();
+    List<UserRecentWorkspace> userRecentWorkspaces =
+        userRecentWorkspaceDao.findByUserIdOrderByLastAccessDateDesc(userId);
+    enforceFirecloudAclsInRecentWorkspaces(userRecentWorkspaces);
+    return userRecentWorkspaceDao.findByUserIdOrderByLastAccessDateDesc(userId);
+  }
+
+  private void enforceFirecloudAclsInRecentWorkspaces(List<UserRecentWorkspace> recentWorkspaces) {
+    List<Workspace> dbWorkspaces =
+        workspaceDao.findAllByWorkspaceIdIn(
+            recentWorkspaces.stream()
+                .map(UserRecentWorkspace::getWorkspaceId)
+                .collect(Collectors.toList()));
+
+    final String email = userProvider.get().getEmail();
+    ImmutableList<Long> idsToDelete =
+        dbWorkspaces.stream()
+            .filter(workspace -> userHasWorkspaceAccess(email, workspace))
+            .map(Workspace::getWorkspaceId)
+            .collect(ImmutableList.toImmutableList());
+
+    if (!idsToDelete.isEmpty()) {
+      userRecentWorkspaceDao.deleteByWorkspaceIdIn(idsToDelete);
+    }
+  }
+
+  private boolean userHasWorkspaceAccess(String email, Workspace workspace) {
+    Map<String, WorkspaceAccessEntry> aclsByEmail =
+        getFirecloudWorkspaceAcls(workspace.getWorkspaceNamespace(), workspace.getFirecloudName());
+    return !aclsByEmail.containsKey(email)
+        || aclsByEmail
+            .get(email)
+            .getAccessLevel()
+            .equals(WorkspaceAccessLevel.NO_ACCESS.toString());
+  }
+
+  @Override
+  public UserRecentWorkspace updateRecentWorkspaces(
+      long workspaceId, long userId, Timestamp lastAccessDate) {
+    Optional<UserRecentWorkspace> maybeRecentWorkspace =
+        userRecentWorkspaceDao.findFirstByWorkspaceIdAndUserId(workspaceId, userId);
+    final UserRecentWorkspace matchingRecentWorkspace =
+        maybeRecentWorkspace
+            .map(
+                recentWorkspace ->
+                    new UserRecentWorkspace(
```

Comment:
I'm not sure what you're asking. The point of looking up the row to begin with is to make sure we're not updating instead of duplicating entries in `user_recent_workspace` if, say, you're flipping back and forth between two workspaces for some reason

Reply:
I understand what this was trying to do, I'm asserting it doesn't work. As written currently there's effectively no difference between the if/else cases of this optional handling.

Reply:
Gotcha - writing a test to cover the flip flopping case now.